lib/wallaby/browser.ex

```diff
@@ -0,0 +1,831 @@
+
+  @doc """
+  Attempts to synchronize with the browser. This is most often used to
+  execute queries repeatedly until it either exceeds the time limit or
+  returns a success.
+
+  ## Note
+
+  It is possible that this function never halts. Whenever we experience a stale
+  reference error we retry the query without checking to see if we've run over
+  our time. In practice we should eventually be able to query the dom in a stable
+  state. However, if this error does continue to occur it will cause wallaby to
+  loop forever (or until the test is killed by exunit).
+  """
+  @opaque result :: {:ok, any()} | {:error, any()}
+  @spec retry((() -> result), timeout) :: result()
+
+  def retry(f, start_time \\ current_time) do
+    case f.() do
+      {:ok, result} ->
+        {:ok, result}
+      {:error, :stale_reference} ->
+        retry(f, start_time)
+      {:error, e} ->
+        if max_time_exceeded?(start_time) do
+          {:error, e}
+        else
+          retry(f, start_time)
+        end
+    end
+  end
+
+  @doc """
+  Fills in a "fillable" element with text. Input elements are looked up by id, label text,
+  or name.
+  """
+  @spec fill_in(element, opts) :: element
+  @spec fill_in(parent, StatelessQuery.t, with: String.t) :: parent
+  @spec fill_in(parent, locator, opts) :: parent
+
+  def fill_in(parent, locator, [{:with, value} | _]=opts) when is_binary(locator) do
+    parent
+    |> find(StatelessQuery.fillable_field(locator, opts))
+    |> fill_in(with: value)
+
+    parent
+  end
+  def fill_in(parent, query, with: value) do
+    parent
+    |> find(query)
+    |> fill_in(with: value)
+
+    parent
+  end
+  def fill_in(%Element{}=element, with: value) when is_number(value) do
+    fill_in(element, with: to_string(value))
+  end
+  def fill_in(%Element{}=element, with: value) when is_binary(value) do
+    element
+    |> clear
+    |> set_value(value)
+
+    element
+  end
+
+  @doc """
+  Chooses a radio button based on id, label text, or name.
+  """
+  @spec choose(Element.t) :: Element.t
+  @spec choose(parent, StatelessQuery.t) :: parent
+  @spec choose(parent, locator, opts) :: parent
+
+  def choose(parent, locator, opts) when is_binary(locator) do
+    parent
+    |> find(StatelessQuery.radio_button(locator, opts))
+    |> click
+
+    parent
+  end
+  def choose(parent, locator) when is_binary(locator) do
+    parent
+    |> find(StatelessQuery.radio_button(locator, []))
+    |> click
+
+    parent
+  end
+  def choose(parent, query) do
+    parent
+    |> find(query)
+    |> click
+
+    parent
+  end
+  def choose(%Element{}=element) do
+    click(element)
+  end
+
+  @doc """
+  Checks a checkbox based on id, label text, or name.
+  """
+  @spec check(Element.t) :: Element.t
+  @spec check(parent, StatelessQuery.t) :: parent
+  @spec check(parent, locator, opts) :: parent
+
+  def check(%Element{}=element) do
+    cond do
+      checked?(element) -> element
+      true -> click(element)
+    end
```

Comment:
I tend to use `cond` over case if the function is going to return a boolean. I don't use `if/else` just in general as `cond` achieves the same thing and is easier to extend down the road.